// pangea-node-sdk/dist/mjs/errors.js
export var PangeaErrors;
(function (PangeaErrors) {
    class PangeaError extends Error {
        constructor(message) {
            super(message);
            this.name = "PangeaGeneralError";
        }
    }
    PangeaErrors.PangeaError = PangeaError;
    class AuditError extends PangeaError {
        constructor(message) {
            super(message);
            this.name = "PangeaAuditError";
        }
    }
    PangeaErrors.AuditError = AuditError;
    class AuditEventError extends AuditError {
        envelope;
        constructor(message, envelope) {
            super(message);
            this.name = "PangeaAuditEventError";
            this.envelope = envelope;
        }
    }
    PangeaErrors.AuditEventError = AuditEventError;
    class APIError extends Error {
        response;
        constructor(message, response) {
            super(message);
            this.name = "PangeanAPIError";
            response.result = response.result;
            this.response = response;
        }
        get pangeaResponse() {
            return this.response;
        }
        get summary() {
            return this.response.summary;
        }
        get errors() {
            return this.response.result?.errors || [];
        }
    }
    PangeaErrors.APIError = APIError;
    //Pangea Validation Errors denoting issues with an API request
    class ValidationError extends APIError {
        constructor(message, response) {
            super(message, response);
            this.name = "ValidationError";
        }
    }
    PangeaErrors.ValidationError = ValidationError;
    //Too many requests were made
    class RateLimiteError extends APIError {
        constructor(message, response) {
            super(message, response);
            this.name = "RateLimiteError";
        }
    }
    PangeaErrors.RateLimiteError = RateLimiteError;
    //API usage requires payment"""
    class NoCreditError extends APIError {
        constructor(message, response) {
            super(message, response);
            this.name = "NoCreditError";
        }
    }
    PangeaErrors.NoCreditError = NoCreditError;
    //User is not authorized to access a given resource
    class UnauthorizedError extends APIError {
        constructor(serviceName, response) {
            let message = "User is not authorized to access service " + serviceName;
            super(message, response);
            this.name = "UnauthorizedError";
        }
    }
    PangeaErrors.UnauthorizedError = UnauthorizedError;
    // Service not enabled
    class ServiceNotEnabledError extends APIError {
        constructor(serviceName, response) {
            let message = String(serviceName) +
                " is not enabled. Go to console.pangea.cloud/service/{service_name} to enable";
            super(message, response);
            this.name = "ServiceNotEnabledError";
        }
    }
    PangeaErrors.ServiceNotEnabledError = ServiceNotEnabledError;
    // No config ID was provided in either token scopes or explicitly
    class MissingConfigID extends APIError {
        constructor(serviceName, response) {
            let message = "Token did not contain a config scope for service " +
                serviceName +
                ". Create a new token or provide a config ID explicitly in the service base";
            super(message, response);
            this.name = "ServiceNotEnabledError";
        }
    }
    PangeaErrors.MissingConfigID = MissingConfigID;
    // Downstream provider error
    class ProviderError extends APIError {
        constructor(message, response) {
            super(message, response);
            this.name = "ProviderError";
        }
    }
    PangeaErrors.ProviderError = ProviderError;
    // A pangea service error
    class InternalServiceErrorError extends APIError {
        constructor(message, response) {
            super(message, response);
            this.name = "InternalServiceError";
        }
    }
    PangeaErrors.InternalServiceErrorError = InternalServiceErrorError;
    // Service is not currently available
    class ServiceNotAvailableError extends APIError {
        constructor(message, response) {
            super(message, response);
            this.name = "ServiceNotAvailableError";
        }
    }
    PangeaErrors.ServiceNotAvailableError = ServiceNotAvailableError;
    class InvalidPayloadReceived extends APIError {
        constructor(message, response) {
            super(message, response);
            this.name = "InvalidPayloadReceived";
        }
    }
    PangeaErrors.InvalidPayloadReceived = InvalidPayloadReceived;
})(PangeaErrors || (PangeaErrors = {}));
