// node-pangea/dist/cjs/services/audit.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_js_1 = __importDefault(require("./base.js"));
const arweave_js_1 = require("../utils/arweave.js");
const verification_js_1 = require("../utils/verification.js");
const signer_js_1 = require("../utils/signer.js");
const utils_js_1 = require("../utils/utils.js");
const errors_js_1 = require("../errors.js");
const SupportedFields = ["actor", "action", "status", "source", "target", "timestamp"];
const SupportedJSONFields = ["message", "new", "old"];
/**
 * AuditService class provides methods for interacting with the Audit Service
 * @extends BaseService
 */
class AuditService extends base_js_1.default {
    constructor(token, config) {
        super("audit", token, config);
        this.publishedRoots = {};
        this.configIdHeaderName = "X-Pangea-Audit-Config-ID";
        this.publishedRoots = {};
        this.apiVersion = "v1";
        this.init();
    }
    /**
     * @summary Log an entry
     * @description Create a log entry in the Secure Audit Log.
     * @param {Object} content - A structured event describing an auditable activity. Supported fields are:
     *   - actor (string): Record who performed the auditable activity.
     *   - action (string): The auditable action that occurred.
     *   - status (string): Record whether or not the activity was successful.
     *   - source (string): Used to record the location from where an activity occurred.
     *   - target (string): Used to record the specific record that was targeted by the auditable activity.
     *   - message (string|object): A message describing a detailed account of what happened.
     *     This can be recorded as free-form text or as a JSON-formatted string.
     *   - new (string|object): The value of a record after it was changed.
     *   - old (string|object): The value of a record before it was changed.
     * @param {Object} options - Log options. The following log options are supported:
     *   - verbose (bool): Return a verbose response, including the canonical event hash and received_at time.
     * @returns {Promise} - A promise representing an async call to the log endpoint.
     * @example
     * ```js
     * const auditData = {
     *    action: "add_employee",
     *    actor: user,
     *    target: data.email,
     *    status: "error",
     *    message: `Resume denied - sanctioned country from ${clientIp}`,
     *    source: "web",
     *  };
     *
     *  const logResponse = await audit.log(auditData);
     * ```
     */
    log(content, options = {}) {
        const event = {
            message: "",
        };
        SupportedFields.forEach((key) => {
            if (key in content) {
                // @ts-ignore
                event[key] = content[key];
            }
        });
        SupportedJSONFields.forEach((key) => {
            if (key in content) {
                // @ts-ignore
                event[key] =
                    // @ts-ignore
                    content[key] instanceof Object
                        ? // @ts-ignore
                            JSON.stringify(content[key])
                        : // @ts-ignore
                            (event[key] = content[key]);
            }
        });
        const data = { event: event };
        if (options.signer) {
            const signer = options.signer;
            const eventJson = (0, utils_js_1.canonicalize)(event);
            const signature = signer.sign(eventJson);
            const pubKey = signer.getPublicKey();
            data.signature = signature;
            data.public_key = pubKey;
        }
        if (options === null || options === void 0 ? void 0 : options.return_hash) {
            data.return_hash = options.return_hash;
        }
        if (options === null || options === void 0 ? void 0 : options.verbose) {
            data.verbose = options.verbose;
        }
        return this.post("log", data);
    }
    /**
     * @summary Search for events
     * @description Search for events that match the provided search criteria.
     * @param {String} query - Natural search string; list of keywords with optional
     *   `<option>:<value>` qualifiers. The following optional qualifiers are supported:
     *   - action:
     *   - actor:
     *   - message:
     *   - new:
     *   - old:
     *   - status:
     *   - target:
     * @param {Object} options - Search options. The following search options are supported:
     *   - limit (number): Maximum number of records to return per page.
     *   - start (string): The start of the time range to perform the search on.
     *   - end (string): The end of the time range to perform the search on. All records up to the latest if left out.
     *   - sources (array): A list of sources that the search can apply to. If empty or not provided, matches only the default source.
     * @returns {Promise} - A promise representing an async call to the search endpoint
     * @example
     * ```js
     * const response = await audit.search("add_employee:Gumby")
     * ```
     */
    search(query, queryOptions, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const defaults = {
                limit: 20,
                order: "desc",
                order_by: "received_at",
                include_membership_proof: false,
                include_hash: false,
                include_root: false,
            };
            const payload = { query };
            Object.assign(payload, defaults);
            Object.assign(payload, queryOptions);
            options.skipEventVerification =
                options.skipEventVerification === undefined ? true : options.skipEventVerification;
            // This would be removed soon.
            payload.include_hash = true;
            if (options === null || options === void 0 ? void 0 : options.verifyConsistency) {
                payload.include_membership_proof = true;
                payload.include_root = true;
            }
            const response = yield this.post("search", payload);
            return this.processSearchResponse(response, options);
        });
    }
    /**
     * @summary Results of a search
     * @description Fetch paginated results of a previously executed search
     * @param {String} id - The id of a successful search
     * @param {number} limit (default 20) - The number of results returned
     * @param {number} offset (default 0) - The starting position of the first returned result
     * @param {boolean} verifyResponse (default false) - Verify consistency and membership proof of every record
     * @returns {Promise} - A promise representing an async call to the results endpoint
     * @example
     * ```js
     * const response = await audit.results(pxx_asd0987asdas89a8, 50, 100)
     * ```
     */
    results(id, limit = 20, offset = 0, options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!id) {
                throw new Error("Missing required `id` parameter");
            }
            const payload = {
                id,
                limit,
                offset,
            };
            options.skipEventVerification =
                options.skipEventVerification === undefined ? true : options.skipEventVerification;
            const response = yield this.post("results", payload);
            return this.processSearchResponse(response, options);
        });
    }
    /**
     * @summary Retrieve tamperproof verification
     * @description Returns current root hash and consistency proof
     * @param {number} size - The size of the tree (the number of records)
     * @returns {Promise} - A promise representing an async call to the results endpoint
     * @example
     * ```js
     * const response = audit.root(7);
     * ```
     */
    root(size = 0) {
        const data = {};
        if (size > 0) {
            data.tree_size = size;
        }
        return this.post("root", data);
    }
    processSearchResponse(response, options) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            if (!response.success) {
                return response;
            }
            const root = response.result.root;
            const localRoot = (treeSize) => __awaiter(this, void 0, void 0, function* () {
                const response = yield this.root(treeSize);
                const root = response.result.data;
                return root;
            });
            if (options === null || options === void 0 ? void 0 : options.skipEventVerification) {
                const v = new signer_js_1.Verifier();
                response.result.events.forEach((record) => {
                    var _a, _b;
                    if (!(0, verification_js_1.verifyLogHash)(record.envelope, record.hash)) {
                        throw new errors_js_1.PangeaErrors.AuditEventError("Error: Fail event hash verification. Hash: " + record.hash, record.envelope);
                    }
                    if (((_a = record.envelope) === null || _a === void 0 ? void 0 : _a.signature) !== undefined && ((_b = record.envelope) === null || _b === void 0 ? void 0 : _b.public_key) !== undefined) {
                        var data = (0, utils_js_1.canonicalize)(record.envelope.event);
                        if (v.verify(data, record.envelope.signature, record.envelope.public_key)) {
                            record.signature_verification = "pass";
                        }
                        else {
                            throw new errors_js_1.PangeaErrors.AuditEventError("Error: Fail event signature verification. Hash: " + record.hash, record.envelope);
                        }
                    }
                    else {
                        record.signature_verification = "none";
                    }
                });
            }
            if (!root) {
                return response;
            }
            if (options === null || options === void 0 ? void 0 : options.verifyConsistency) {
                if (!(root === null || root === void 0 ? void 0 : root.tree_name))
                    return response;
                const treeName = root === null || root === void 0 ? void 0 : root.tree_name;
                const treeSizes = new Set();
                treeSizes.add((_a = root === null || root === void 0 ? void 0 : root.size) !== null && _a !== void 0 ? _a : 0);
                response.result.events.forEach((record) => {
                    if (record.leaf_index !== undefined) {
                        const idx = Number(record.leaf_index);
                        treeSizes.add(idx + 1);
                        if (idx > 0) {
                            treeSizes.add(idx);
                        }
                    }
                });
                this.publishedRoots = yield (0, arweave_js_1.getArweavePublishedRoots)(treeName, Array.from(treeSizes), localRoot);
                response.result.events.forEach((record) => {
                    if (record.leaf_index) {
                        const consistency = (0, verification_js_1.verifyConsistencyProof)({
                            publishedRoots: this.publishedRoots,
                            record: record,
                        });
                        // TODO: Should we do a Enum for this? Or maybe true, false, undefined?
                        record.consistency_verification = consistency ? "pass" : "fail";
                    }
                    else {
                        record.consistency_verification = "none";
                    }
                    if (record.membership_proof) {
                        const membership = (0, verification_js_1.verifyMembershipProof)({
                            root: response.result.root,
                            record: record,
                        });
                        record.membership_verification = membership ? "pass" : "fail";
                    }
                    else {
                        record.membership_verification = "none";
                    }
                });
            }
            return response;
        });
    }
}
exports.default = AuditService;
