// pangea-node-sdk/dist/cjs/client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_js_1 = __importDefault(require("./config.js"));
const audit_js_1 = __importDefault(require("./services/audit.js"));
const embargo_js_1 = __importDefault(require("./services/embargo.js"));
const redact_js_1 = __importDefault(require("./services/redact.js"));
class PangeaClient {
    constructor(token, config) {
        this.token = token;
        this.config = config || new config_js_1.default();
        // service references
        this.auditService = null;
        this.redactService = null;
        this.embargoService = null;
    }
    // TODO: auto-generate these wrappers
    audit(token, config) {
        const auditToken = token || this.token;
        const auditConfig = config || this.config;
        if (!this.auditService) {
            this.auditService = new audit_js_1.default(auditToken, auditConfig);
        }
        return this.auditService;
    }
    redact(token, config) {
        const redactToken = token || this.token;
        const redactConfig = config || this.config;
        if (!this.redactService) {
            this.redactService = new redact_js_1.default(redactToken, redactConfig);
        }
        return this.redactService;
    }
    embargo(token, config) {
        const embargoToken = token || this.token;
        const embargoConfig = config || this.config;
        if (!this.embargoService) {
            this.embargoService = new embargo_js_1.default(embargoToken, embargoConfig);
        }
        return this.embargoService;
    }
}
exports.default = PangeaClient;
