// node-pangea/dist/cjs/response.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PangeaResponse = exports.ResponseObject = void 0;
const SupportedJSONFields = ["message", "new", "old"];
/**
 * Pangea Response object
 */
class ResponseObject {
    constructor(body) {
        this.request_id = "InvalidPayloadReceived";
        this.request_time = "InvalidPayloadReceived";
        this.response_time = "InvalidPayloadReceived";
        this.status = "InvalidPayloadReceived";
        this.result = {};
        this.summary = "InvalidPayloadReceived";
        Object.assign(this, body);
    }
}
exports.ResponseObject = ResponseObject;
class PangeaResponse extends ResponseObject {
    constructor(response) {
        const obj = JSON.parse(JSON.stringify(response.body), parseJSONfields);
        super(obj);
        this.gotResponse = response;
        this.success = this.status === "Success";
    }
}
exports.PangeaResponse = PangeaResponse;
function parseJSONfields(key, value) {
    if (SupportedJSONFields.includes(key) && typeof value === "string") {
        try {
            // @ts-ignore
            const obj = JSON.parse(value);
            return obj;
        }
        catch (e) {
            return value;
        }
    }
    else {
        return value;
    }
}
exports.default = PangeaResponse;
